[pulseaudio] Declare pulseaudio module per audio stream

GiviMAD opened the pull request:
<s>These PR depends on https://github.com/openhab/openhab-addons/pull/16111 but I'm open the PR for visibility, I will remove the WIP after that one gets merged and rebase the PR.</s>

Hello, I would like to merge these changes to the pulseaudio binding, I did those while Christmas and have been using them for a couple of weeks, seems to work correctly.

The reasons for the changes are:

* Declare the sink modules in stream format instead of assume the default pulse audio config (previously the binding assumed that the pulseaudio format is 44100hz 16bit 2 channels, which is the default but can be modified, causing the binding to work incorrectly if you have modified those), these also avoid to convert the audio for pcm streams using the Java audio api as those are already supported by pulseaudio.
* Be able to play two audio stream at the same time. The previous implementation of the sink assumed that you can send concurrent audio streams using a single socket connection to pulseaudio, that is not correct, you should open several sockets to transmit the audio and pulse will mix them correctly and play them concurrently.
* I have fixed some minor issues that I found in the meanwhile, like disposing the source on consecutive errors, also the sink/source instances were not correctly assigned to the class properties, causing them to be recreated over and over again on the status check, which I think was causing some memory issues for me.

I have removed some options, like the preferred port, as I think it does not make sense any more, because now the binding creates and disposes the modules, and will not use modules that has been not created by it.

Hope everything is correct.
Best Regards!

GiviMAD commented:
@dalgwen if you have any chance to review these PR I'll appreciate any feedback you can give me.

dalgwen commented on bundles/org.openhab.binding.pulseaudio/src/main/java/org/openhab/binding/pulseaudio/internal/PulseAudioAudioSink.java:
```diff
@@ -74,69 +78,123 @@ public void process(@Nullable AudioStream audioStream)
         }
-        addClientCount();
-        try (ConvertedInputStream normalizedPCMStream = new ConvertedInputStream(audioStream)) {
-            for (int countAttempt = 1; countAttempt <= 2; countAttempt++) { // two attempts allowed
```
I put this retry mechanism because I sometimes encountered "zombie connections" : They are connections that can't be seen as failed unless we try to write to them. We can encouter this especially if we use the value -1 for timeout (keeping connection).

So, I see two options : 
- keeping it
- or removing it, and also removing the ability to keep connection after playing (which is fine for me if wanted, since latency (even when reconnecting, which was the reason of this parameter) seems to have improved to absolutely not being noticeable in my last test)

GiviMAD replied:
Thank you for the review @dalgwen! 

While developing I thought on the possibility of removing it (the "keep connection" functionality), but I wanted to know first whetter you found a big difference with it because in my case it's also not noticeable, thank you for testing it. I'm going to remove it then.

dalgwen commented on bundles/org.openhab.binding.pulseaudio/src/main/java/org/openhab/binding/pulseaudio/internal/PulseAudioAudioSink.java:
```diff
@@ -74,69 +78,123 @@ public void process(@Nullable AudioStream audioStream)
+            }
+        }
+        Optional<SimpleProtocolTCPModule> spModule = Optional.empty();
+        try (pcmSignedStream) {
```
Isn't there an issue when playing PipedAudioStream ?

With a try-with, the Closeable resource is always closed at the end of the block.
But, in the case of a PipedAudioStream, the sound is played asynchronously, in another thread. So the try-with block will end, without considering the asyncStream.
In this case, wouldn't the stream be closed to soon ?

GiviMAD replied:
Yes, it was wrong, I have made some refactor/fixes in the processAndComplete sink method. I think that now the audio stream and the module are closed/released in the correct moment.

dalgwen commented on bundles/org.openhab.binding.pulseaudio/src/main/java/org/openhab/binding/pulseaudio/internal/PulseaudioSimpleProtocolStream.java:
```diff
@@ -32,86 +40,198 @@
+
     /**
      * Connect to pulseaudio with the simple protocol
      * Will schedule an attempt for disconnection after timeout
```
Is is still true ?
From what I read, disconnection won't happen, unless some sound end playing.

GiviMAD replied:
I've made some refactor in this method to return a container class that contains the module and runnable instance that allows to release it. I think it makes the code in the consumers easy to write/read. 

After the changes in the processAndComplete sink method, I think the module is correctly released after usage. Thanks for notice it.

dalgwen commented:
Sorry for the delay !

> I have removed some options, like the preferred port

I'm personnaly OK with that, because I don't use it, but did you think about firewall issue ?
I put this option because I think some people use firewall, even on their LAN. So the ability to choose (fix) a port number can be mandatory for them. But TBH, I'm not even sure it is used, so I totally support the decision if the use case is deemed too narrow to keep this. Otherwise, maybe we can imagine a configuration parameter for a plage of ports to try when loading a module ?

dalgwen reviewed (commented):
I tested my use cases and it works.
Also, playing sound concurrently is now OK, thanks !

Disclaimer : I didn't read the new code extensively ; I mainly focused on legacy code.
But I can at least tell it works (again, with my simple use cases)

GiviMAD commented:
> Sorry for the delay !
> 
> > I have removed some options, like the preferred port
> 
> I'm personnaly OK with that, because I don't use it, but did you think about firewall issue ? I put this option because I think some people use firewall, even on their LAN. So the ability to choose (fix) a port number can be mandatory for them. But TBH, I'm not even sure it is used, so I totally support the decision if the use case is deemed too narrow to keep this. Otherwise, maybe we can imagine a configuration parameter for a plage of ports to try when loading a module ?

Thank you, I didn't realize that. And yes, I think adding a min port number and max port number as configuration parameters can be enough to handle that situation, and it's not much change, it's worth to added it.

GiviMAD commented on bundles/org.openhab.binding.pulseaudio/src/main/java/org/openhab/binding/pulseaudio/internal/PulseAudioAudioSink.java:
```diff
@@ -74,69 +78,123 @@ public void process(@Nullable AudioStream audioStream)
+            }
+        } else {
+            try {
+                var convertedInputStream = new ConvertedInputStream(audioStream);
```
Inside that try block the only line that can throw an exception is that one, so there is nothing to release in the catch block if the exception is thrown. Then it's assigned to variable `pcmSignedStream` and after the last changes it should be correctly closed after usage or in case of errors. I think it's consistent now.

dalgwen commented:
Oops, @GiviMAD , I accidentally pushed  to your repository. I didn't even though it was possible !?? Sorry for the mistake.
I was trying to push to mine, because I have some proposal for this PR.

I made a deeper review this time, and, as I was reading the PulseAudioAudioSink class, I thought of improving the readability a bit.
Instead of bothering you with details, I tried to do it myself. Can you please code review my code review 😅 and maybe include it in your pull request ? Feel free to hard reset back, as me pushing to your branch was clearly a mistake.

This refactor is based on some assomptions, so if one is wrong I will of course have to modify or even retract my proposal.
List of changes :
- Moved some logic of stream preparation and parsing inside ConvertedInputStream. ConvertedInputStream is now responsible of "preparing and cleaning the stream" (for example removing FMT). The goal is to pass through ConvertedInputStream to prepare the stream every time, even if we do nothing (I want to improve readability by removing branches in PulseAudioAudioSink). Maybe we can then rename ConvertedInputStream ?
- removed the second conversion pass in ConvertedInputStream, it is no longer necessary as you broadened the possibility of the simple-tcp-module by enabling different PCM source format.
- removing all alternates way of computing sound duration and keeping only the last one : I realize that, as the sound is always converted to PCM, and that we know the bitrate and can find the amount transfered (by waiting after the transfer), we have the information to make a precise computation in every case, no matter the source format. I also find a bug and made a fix [to core for this](https://github.com/openhab/openhab-core/pull/4116). I also keep the sound duration computation, even with the PipedAudioStream : it occurs at the end, with no overhead, and it simplifies thing by removing the need to branch.
By the way, I think branching depending on the audio stream class is a bad design we should avoid. It seems a violation of the Interface contract and it should not be the Sink responsibility to handle differences between AudioInputStream ?
- So, no need to have a branch for PipedStream / other stream ? We can handle both of them the same way : asynchronously.
While removing this if branch, I found two differences in your code between PipedStream and other stream that I couldn't explain, and I didn't keep them, so I hope it won't affect the behaviour : 
	-- I removed the canceled(true) in the onClose callback. When reading from pipedAudioStream, if the stream is closed, you set cancel with a callback. But cancel doesn't seem necessary to stop the loop, because when it's closed reading return -1 and it will also stop the loop. I hope I was right ?
	-- When in the branch to play PipedAudioStream, there is no disconnection on IOException. But there is one in the branch for the other kind of AudioStream. I assumed it was a mistake and put it back in the mutualized code.
- I factorised error/closing/releasing management in a "endStream" method
- I fixed an error with the port range in PulseAudio Client

dalgwen commented:
I have another comment : on PipedAudioStream, you have a "closed" AtomicBoolean.
Is AtomicBoolean really mandatory ? Manipulating this class involves locking on every read / change.
A lock is a costly operation, and this AtomicBoolean is everywhere, on every read, even for a single byte.

dalgwen commented:
And a last one, about the simple-tcp-module : when stopping openHAB, the module(s) stay loaded on the pulseaudio server side.
It is not an issue, but when openHAB restarts, it doesn't read this module as his, and it load another ones.
So during my test, I stop and started openHAB several times and simple-tcp-modules accumulated on the pulseaudio server.
Do you think you can do something about it ?

dalgwen commented:
I see that I didn't include a link to the merge request to you repo.
And that I was targeting the wrong branch. I fixed that.
I shouldn't do code at 2AM  😅
So here is [the link](https://github.com/GiviMAD/openhab-addons/pull/3).

GiviMAD commented:
> I have another comment : on PipedAudioStream, you have a "closed" AtomicBoolean. Is AtomicBoolean really mandatory ? Manipulating this class involves locking on every read / change. A lock is a costly operation, and this AtomicBoolean is everywhere, on every read, even for a single byte.

My opinion is that the boolean should be atomic because it's modified/accessed from different threads. But that will be something to report in the core, not part of these PR.



> And a last one, about the simple-tcp-module : when stopping openHAB, the module(s) stay loaded on the pulseaudio server side. It is not an issue, but when openHAB restarts, it doesn't recognize this module as its, and it loads another one. So during my test, I stop and started openHAB several times and simple-tcp-modules accumulated on the pulseaudio server. Do you think you can do something about it ?

Are you running the dialog? I think there could be a race condition there because for me sometimes the sink module keeps loaded after a restart, but only happens for the sink.



> I see that I didn't include a link to the merge request to you repo.
> And that I was targeting the wrong branch. I fixed that.
> I shouldn't do code at 2AM 😅
> So here is [the link](https://github.com/GiviMAD/openhab-addons/pull/3).

I have added a comment there. The PipedAudioStream will be used to send infinite audio streams, meaning the consumer will never reach the end of stream, that is why it includes the "onClose" methods, so consumer can stop reading "gracefully" before the stream gets closed. Also trying to measure its duration is useless you just what to wait for it to get closed. With your PR you are removing that functionality and it's important to me for future development so I'll like to keep it.

dalgwen commented:
> Are you running the dialog? I think there could be a race condition there because for me sometimes the sink module keeps loaded after a restart, but only happens for the sink.

No, I'm not testing it with a dialog.
Steps  : 
- start my openHAB dev env
- count on the remote server the number of tcp-client module already loaded (`pactl list | grep -i module-simple-protocol-tcp | wc -l`). (Result 0)
- play a sound to a remote pulse audio
- count on the remote server the number. (Result 1)
- stop my openHAB dev
- count on the remote server (Result 1)
- Restart openHAB dev
- play a sound
- count on the remote server (Result 2)
- repeat. Every loop adds a module loaded on the remote

I think we could add something like that : 
At start, browsing the module simple-tcp-module-client and add them to the Idle list ?

dalgwen commented:
There is also a logic issue with this case : 
- Already one module (let's called it module1) is used to play something, on port X
- Trying to play another sound
- No idle module remaining, and we call PulseAudioClient.loadModuleSimpleProtocolTcpIfNeeded with spModule parameter = null
- get a random port (line 359)
- Bad luck, the random port is X (there is a chance, especially if the range is small)
- called sendRawCommand to create the module on the server -> no effect, it already exists
- it is found when we check if it is created (of course, as it already exists)
- a new SimpleProtocolTCPModule (let's called it module2) is returned. BUT it is the same a module1, with the same port.
- we can now have two AudioSink that can used the same port at the same time

GiviMAD commented:
Hey @dalgwen sorry for the delay.

> ...
> 
> I think we could add something like that : At start, browsing the matching simple-tcp-module-client and add them to the Idle list of the corresponding sink ?

Yes, I will apply that solution, I was doing some testing the other day trying to understand why the close method is not called (I'm able to replicate the problem by restarting the docker container) but I don't understand what is the problem, so I will apply your suggested solution and add to the idle list all the pre-existent "simple protocol modules" in the configured port range.

> ...
>
> We maybe should have a list of port in use somewhere?

I also thought at these, but as the binding checks for source|sink + format + port combination I saw it as an edge case.
But we already have a list of ports in the module list so I think we can add an extra check at "get a random port (line 359)" so it never return a port in use and prevent the problem.

I'll look at it and ping you. Thank you for the comments.

GiviMAD commented:
@dalgwen, I think that the latest comment are addressed on the last version.

Instead of reusing the pre-existent modules for the device, I opted by removing them to avoid adding more complexity. I also added the check to avoid trying with a port already in use by another simple protocol tcp module instance.

dalgwen commented on bundles/org.openhab.binding.pulseaudio/src/main/java/org/openhab/binding/pulseaudio/internal/PulseaudioClient.java:
```diff
@@ -369,76 +431,80 @@ public Optional<Integer> loadModuleSimpleProtocolTcpIfNeeded(AbstractAudioDevice
-        List<Module> modulesCopy = new ArrayList<>(modules);
-        var isSource = item instanceof Source;
-        return modulesCopy.stream() // iteration on modules
-                .filter(module -> MODULE_SIMPLE_PROTOCOL_TCP_NAME.equals(module.getPaName())) // filter on module name
```
Did you remove ` .filter(module -> MODULE_SIMPLE_PROTOCOL_TCP_NAME.equals(module.getPaName()))  ` on purpose ?
(I'm not sure why it was here...)

GiviMAD replied:
Yes, because now all of those are instances of SimpleProtocolTCPModule, and that is what I'm using to differentiate them from the other module instances.
```java
     ...
     modulesCopy.stream() //
        .filter(SimpleProtocolTCPModule.class::isInstance) //
        .map(SimpleProtocolTCPModule.class::cast) //
        ...
```

dalgwen reviewed (commented):
Apart from this two minor questions/remarks , all good.
(again, sorry for the delay)

lsiepel reviewed (changes requested):
Code LGTM, just two comments to pay attention to.
1. The changes in thing configuration parameters should be documented in the readm.md
2. As the parameters have changed, but a default is provided, i think this is backwards compatible, but i suspect some setups might break. Does this make sense? If so please provide a notice in the distro upgrade.

GiviMAD commented:
> Code LGTM, just two comments to pay attention to.
> 
> 1. The changes in thing configuration parameters should be documented in the readm.md
> 2. As the parameters have changed, but a default is provided, i think this is backwards compatible, but i suspect some setups might break. Does this make sense? If so please provide a notice in the distro upgrade.

Hello @lsiepel, these PR makes some changes on the way the binding interact with pulseaudio for the sink, but it's not a breaking changes.

The major differences are that previously it used a preferred port on the remote host and now it gets a random port but the previous implementation fallback to also using a random port, and also we have removed the option for keeping the ports open because it seems to bring no notable performance speed, so it's unlikely anyone will face problems updating or downgrading (in fact I did both). I don't think it needs to be mentioned in the distro upgrade.

Also the removed parameter were not documented on the README, that is why I didn't document the new ones, but I can add a table with all the parameters as there is in other bindings if you think is a nice to have.

lsiepel commented:
> > Code LGTM, just two comments to pay attention to.
> > 
> > 1. The changes in thing configuration parameters should be documented in the readm.md
> > 2. As the parameters have changed, but a default is provided, i think this is backwards compatible, but i suspect some setups might break. Does this make sense? If so please provide a notice in the distro upgrade.
> 
> Hello @lsiepel, the PR makes some changes on the way the binding interact with pulseaudio for the sink, but it's not a breaking change.
> 
> The major differences are that previously it used a preferred port on the remote host and now it gets a random port but the previous implementation fallback to also using a random port, and also we have removed the option for keeping the ports open because it seems to bring no notable performance speed, so it's unlikely anyone will face problems updating or downgrading (in fact I did both). I don't think it needs to be mentioned in the distro upgrade.
> 
> Also the removed parameter were not documented on the README, that is why I didn't document the new ones, but I can add a table with all the parameters as there is in other bindings if you think is a nice to have.

Ok thanks for the details, yes please document the parameters.

GiviMAD commented:
> > > Code LGTM, just two comments to pay attention to.
> > > 
> > > 1. The changes in thing configuration parameters should be documented in the readm.md
> > > 2. As the parameters have changed, but a default is provided, i think this is backwards compatible, but i suspect some setups might break. Does this make sense? If so please provide a notice in the distro upgrade.
> > 
> > 
> > Hello @lsiepel, the PR makes some changes on the way the binding interact with pulseaudio for the sink, but it's not a breaking change.
> > The major differences are that previously it used a preferred port on the remote host and now it gets a random port but the previous implementation fallback to also using a random port, and also we have removed the option for keeping the ports open because it seems to bring no notable performance speed, so it's unlikely anyone will face problems updating or downgrading (in fact I did both). I don't think it needs to be mentioned in the distro upgrade.
> > Also the removed parameter were not documented on the README, that is why I didn't document the new ones, but I can add a table with all the parameters as there is in other bindings if you think is a nice to have.
> 
> Ok thanks for the details, yes please document the parameters.

I have added the things configuration to the readme.

lsiepel reviewed (approved):
Thanks, LGTM
Edit: Removed the empty line myself to prevent SAT warning.
